plugins/modules/primary_ip.py

```diff
@@ -163,13 +204,20 @@ def _get_primary_ip(self):
             self.fail_json_hcloud(exception)
 
     def _create_primary_ip(self):
-        self.module.fail_on_missing_params(required_params=["type", "datacenter"])
+        self.module.fail_on_missing_params(required_params=["type", "name"])
         try:
             params = {
                 "type": self.module.params.get("type"),
                 "name": self.module.params.get("name"),
-                "datacenter": self.client.datacenters.get_by_name(self.module.params.get("datacenter")),
+                "auto_delete": self.module.params.get("auto_delete"),
+                "datacenter": None,
             }
+            if self.module.params.get("server") is not None:
+                params["assignee_id"] = self._client_get_by_name_or_id("servers", self.module.params.get("server")).id
+            elif self.module.params.get("datacenter") is not None:
+                params["datacenter"] = self.client.datacenters.get_by_name(self.module.params.get("datacenter"))
+            else:
+                self.module.fail_json(msg="server or datacenter is required")
```

Comment:
Maybe using `check_required_one_of` from the `ansible.module_utils.common.validation` module is a better idea here. Not sure if this is already in ansible-core 2.14 though.

Reply:
https://docs.ansible.com/ansible/latest/porting_guides/porting_guide_core_2.11.html

Reply:
Could you provide some input on this?  Not really sure how to implement and cannot find any examples.

Reply:
I have never used it either, but you could try to do add this under the existing validation for the create method:
```py
from ansible.module_utils.common.validation import check_required_one_of
from ansible.module_utils.errors import RequiredOneOfError

#...
    def _create_primary_ip(self):
        self.module.fail_on_missing_params(required_params=["type", "name"])
        try:
            check_required_one_of(["server", "datacenter"], self.module.params)
        except TypeError as exc:
            raise RequiredOneOfError(exc) from exc
```

Haven't tested it though. Maybe reraising the TypeError as RequiredOneOfError is not really useful.

Reply:
Looks like it does not need the try/except, same results.

Without:
`check_required_one_of(["server", "datacenter"], self.module.params)`
```
TASK [test] *******************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:10
fatal: [localhost]: FAILED! => changed=false
  invocation:
    module_args:
      api_endpoint: https://api.hetzner.cloud/v1
      api_token: VALUE_SPECIFIED_IN_NO_LOG_PARAMETER
      auto_delete: false
      datacenter: null
      delete_protection: null
      id: null
      labels: null
      name: primary-ip-test
      server: null
      state: present
      type: ipv4
  msg: server or datacenter is required
...ignoring
Tuesday 20 February 2024  12:13:33 -0500 (0:00:00.841)       0:00:01.646 ******

TASK [debug] ***********************************************************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:18
ok: [localhost] =>
  msg:
    changed: false
    failed: true
    msg: server or datacenter is required
```

With try:

```
try:
            check_required_one_of(["server", "datacenter"], self.module.params)
except TypeError as exc:
            raise RequiredOneOfError(exc) from exc
```
```
TASK [test] *******************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:10
fatal: [localhost]: FAILED! => changed=false
  invocation:
    module_args:
      api_endpoint: https://api.hetzner.cloud/v1
      api_token: VALUE_SPECIFIED_IN_NO_LOG_PARAMETER
      auto_delete: false
      datacenter: null
      delete_protection: null
      id: null
      labels: null
      name: primary-ip-test
      server: null
      state: present
      type: ipv4
  msg: server or datacenter is required
...ignoring
Tuesday 20 February 2024  12:15:08 -0500 (0:00:01.371)       0:00:02.284 ******

TASK [debug] ***********************************************************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:18
ok: [localhost] =>
  msg:
    changed: false
    failed: true
    msg: server or datacenter is required
```

Reply:
Ok that function and most of those validators are dumb.  They check if parameter is present but not set however figured it out.

Reply:
Copied this method:

```python
    def fail_on_missing_params(self, required_params=None):
        if not required_params:
            return
        try:
            check_missing_parameters(self.params, required_params)
        except TypeError as e:
            self.fail_json(msg=to_native(e))
```

Which becomes:
```python
    def _create_primary_ip(self):
        self.module.fail_on_missing_params(required_params=["type", "name"])
        try:
            check_required_one_of([["server", "datacenter"]], {k: v for k, v in self.module.params.items() if v})
        except TypeError as e:
            self.module.fail_json(msg=to_native(e))
```

New test:
```yaml
- name: test no datacenter or server
  hetzner.hcloud.primary_ip:
    name: "{{ hcloud_primary_ip_name }}fail"
    type: ipv4
    state: present
  ignore_errors: true
  register: result
- name: verify test no datacenter or server
  assert:
    that:
      - result is failed
      - 'result.msg == "one of the following is required: server, datacenter"'
```

```
08:47 TASK [server : test no datacenter or server] ***********************************
08:48 fatal: [testhost]: FAILED! => {"changed": false, "msg": "one of the following is required: server, datacenter"}
08:48 ...ignoring
08:48 
08:48 TASK [server : verify test no datacenter or server] ****************************
08:48 ok: [testhost] => {
08:48     "changed": false,
08:48     "msg": "All assertions passed"
08:48 }
```